addons/sale_stock/models/sale_order.py

```diff
@@ -143,10 +143,23 @@ def action_view_delivery(self):
     def action_cancel(self):
         documents = None
         for sale_order in self:
+            if any(picking.state == 'done' for picking in sale_order.picking_ids) and all(i == 0 for i in sale_order.order_line.mapped('qty_delivered')):
```

Comment:
Batch working not supported. This will crash when self is a recordset of multiple orders (on `self.id`).
Either support multiple sale.order ids in the wizard and give them as context, or cancel the order, and its moves, directly on button click.